File: goldencheck/engine/differ.py
"""Diff engine — compare two versions of a data file."""
from __future__ import annotations

from dataclasses import dataclass, field

import polars as pl

from goldencheck.models.finding import Finding, Severity
from goldencheck.models.profile import DatasetProfile


@dataclass
class SchemaChange:
    change_type: str  # "added", "removed", "type_changed"
    column: str
    old_type: str | None = None
    new_type: str | None = None


@dataclass
class FindingChange:
    change_type: str  # "new", "resolved", "worsened", "improved"
    column: str
    check: str
    old_severity: str | None = None
    new_severity: str | None = None
    old_rows: int = 0
    new_rows: int = 0
    message: str = ""


@dataclass
class StatChange:
    metric: str
    old_value: str
    new_value: str
    delta: str


@dataclass
class DiffReport:
    schema_changes: list[SchemaChange] = field(default_factory=list)
    finding_changes: list[FindingChange] = field(default_factory=list)
    stat_changes: list[StatChange] = field(default_factory=list)


def diff_files(
    old_df: pl.DataFrame,
    new_df: pl.DataFrame,
    old_findings: list[Finding],
    new_findings: list[Finding],
    old_profile: DatasetProfile | None = None,
    new_profile: DatasetProfile | None = None,
) -> DiffReport:
    """Compare two versions of a dataset and return a diff report."""
    report = DiffReport()

    # --- Schema changes ---
    old_cols = set(old_df.columns)
    new_cols = set(new_df.columns)

    for col in sorted(new_cols - old_cols):
        report.schema_changes.append(SchemaChange(
            change_type="added", column=col,
            new_type=str(new_df[col].dtype),
        ))

    for col in sorted(old_cols - new_cols):
        report.schema_changes.append(SchemaChange(
            change_type="removed", column=col,
            old_type=str(old_df[col].dtype),
        ))

    for col in sorted(old_cols & new_cols):
        old_type = str(old_df[col].dtype)
        new_type = str(new_df[col].dtype)
        if old_type != new_type:
            report.schema_changes.append(SchemaChange(
                change_type="type_changed", column=col,
                old_type=old_type, new_type=new_type,
            ))

    # --- Finding changes ---
    old_by_key = {}
    for f in old_findings:
        if f.severity >= Severity.WARNING:
            key = (f.column, f.check)
            if key not in old_by_key or f.severity > old_by_key[key].severity:
                old_by_key[key] = f

    new_by_key = {}
    for f in new_findings:
        if f.severity >= Severity.WARNING:
            key = (f.column, f.check)
            if key not in new_by_key or f.severity > new_by_key[key].severity:
                new_by_key[key] = f

    all_keys = set(old_by_key.keys()) | set(new_by_key.keys())
    for key in sorted(all_keys):
        old_f = old_by_key.get(key)
        new_f = new_by_key.get(key)

        if new_f and not old_f:
            report.finding_changes.append(FindingChange(
                change_type="new", column=key[0], check=key[1],
                new_severity=new_f.severity.name,
                new_rows=new_f.affected_rows,
                message=new_f.message,
            ))
        elif old_f and not new_f:
            report.finding_changes.append(FindingChange(
                change_type="resolved", column=key[0], check=key[1],
                old_severity=old_f.severity.name,
                old_rows=old_f.affected_rows,
                message=old_f.message,
            ))
        elif old_f and new_f:
            if new_f.severity > old_f.severity or new_f.affected_rows > old_f.affected_rows * 1.5:
                report.finding_changes.append(FindingChange(
                    change_type="worsened", column=key[0], check=key[1],
                    old_severity=old_f.severity.name, new_severity=new_f.severity.name,
                    old_rows=old_f.affected_rows, new_rows=new_f.affected_rows,
                    message=new_f.message,
                ))
            elif new_f.severity < old_f.severity or new_f.affected_rows < old_f.affected_rows * 0.5:
                report.finding_changes.append(FindingChange(
                    change_type="improved", column=key[0], check=key[1],
                    old_severity=old_f.severity.name, new_severity=new_f.severity.name,
                    old_rows=old_f.affected_rows, new_rows=new_f.affected_rows,
                    message=new_f.message,
                ))

    # --- Stat changes ---
    old_rows = len(old_df)
    new_rows = len(new_df)
    if old_rows != new_rows:
        pct = ((new_rows - old_rows) / old_rows * 100) if old_rows > 0 else 0
        report.stat_changes.append(StatChange(
            metric="Rows",
            old_value=f"{old_rows:,}",
            new_value=f"{new_rows:,}",
            delta=f"{pct:+.0f}%",
        ))

    if len(old_df.columns) != len(new_df.columns):
        diff = len(new_df.columns) - len(old_df.columns)
        report.stat_changes.append(StatChange(
            metric="Columns",
            old_value=str(len(old_df.columns)),
            new_value=str(len(new_df.columns)),
            delta=f"{diff:+d}",
        ))

    return report


def format_diff_report(report: DiffReport, label: str = "") -> str:
    """Format a DiffReport as human-readable text."""
    lines = []
    if label:
        lines.append(f"goldencheck diff — {label}")
        lines.append("")

    if report.schema_changes:
        lines.append("Schema changes:")
        for c in report.schema_changes:
            if c.change_type == "added":
                lines.append(f"  + {c.column} ({c.new_type})")
            elif c.change_type == "removed":
                lines.append(f"  - {c.column} ({c.old_type})")
            elif c.change_type == "type_changed":
                lines.append(f"  ~ {c.column}: {c.old_type} -> {c.new_type}")
        lines.append("")

    if report.finding_changes:
        lines.append("Finding changes:")
        for c in report.finding_changes:
            if c.change_type == "new":
                lines.append(f"  NEW   [{c.column}] {c.message[:60]}")
            elif c.change_type == "resolved":
                lines.append(f"  FIXED [{c.column}] {c.check} resolved")
            elif c.change_type == "worsened":
                lines.append(f"  WORSE [{c.column}] {c.old_rows} -> {c.new_rows} rows")
            elif c.change_type == "improved":
                lines.append(f"  BETTER [{c.column}] {c.old_rows} -> {c.new_rows} rows")
        lines.append("")

    if report.stat_changes:
        lines.append("Stats:")
        for c in report.stat_changes:
            lines.append(f"  {c.metric}: {c.old_value} -> {c.new_value} ({c.delta})")
        lines.append("")

    if not report.schema_changes and not report.finding_changes and not report.stat_changes:
        lines.append("No changes detected.")

    return "\n".join(lines)

File: packages/goldencheck-js/src/core/engine/differ.ts
/**
 * Differ — compares two dataset versions.
 * Port of goldencheck/engine/differ.py.
 */

import type { TabularData } from "../data.js";
import type { Finding, DatasetProfile } from "../types.js";
import { Severity } from "../types.js";

export interface SchemaChange {
  readonly type: "added" | "removed" | "type_changed";
  readonly column: string;
  readonly oldType?: string | undefined;
  readonly newType?: string | undefined;
}

export interface FindingChange {
  readonly type: "new" | "resolved" | "worsened" | "improved";
  readonly column: string;
  readonly check: string;
  readonly oldSeverity?: string | undefined;
  readonly newSeverity?: string | undefined;
  readonly message: string;
}

export interface StatChange {
  readonly metric: string;
  readonly oldValue: number;
  readonly newValue: number;
  readonly delta: number;
}

export interface DiffReport {
  readonly schemaChanges: readonly SchemaChange[];
  readonly findingChanges: readonly FindingChange[];
  readonly statChanges: readonly StatChange[];
}

/**
 * Compare two datasets and their findings.
 */
export function diffData(
  oldData: TabularData,
  newData: TabularData,
  oldFindings: readonly Finding[],
  newFindings: readonly Finding[],
): DiffReport {
  const schemaChanges: SchemaChange[] = [];
  const findingChanges: FindingChange[] = [];
  const statChanges: StatChange[] = [];

  // Schema changes
  const oldCols = new Set(oldData.columns);
  const newCols = new Set(newData.columns);

  for (const col of newCols) {
    if (!oldCols.has(col)) {
      schemaChanges.push({ type: "added", column: col, newType: newData.dtype(col) });
    }
  }
  for (const col of oldCols) {
    if (!newCols.has(col)) {
      schemaChanges.push({ type: "removed", column: col, oldType: oldData.dtype(col) });
    }
  }
  for (const col of oldCols) {
    if (newCols.has(col)) {
      const oldType = oldData.dtype(col);
      const newType = newData.dtype(col);
      if (oldType !== newType) {
        schemaChanges.push({ type: "type_changed", column: col, oldType, newType });
      }
    }
  }

  // Finding changes — match by (column, check), supporting multiple per key
  const oldGroups = new Map<string, Finding[]>();
  for (const f of oldFindings) {
    const key = `${f.column}|${f.check}`;
    const arr = oldGroups.get(key);
    if (arr) arr.push(f);
    else oldGroups.set(key, [f]);
  }
  const newGroups = new Map<string, Finding[]>();
  for (const f of newFindings) {
    const key = `${f.column}|${f.check}`;
    const arr = newGroups.get(key);
    if (arr) arr.push(f);
    else newGroups.set(key, [f]);
  }

  for (const [key, newFs] of newGroups) {
    const oldFs = oldGroups.get(key);
    if (!oldFs) {
      // All new
      for (const newF of newFs) {
        findingChanges.push({
          type: "new",
          column: newF.column,
          check: newF.check,
          newSeverity: severityName(newF.severity),
          message: newF.message,
        });
      }
    } else {
      // Compare by index (best-effort pairing)
      const maxLen = Math.max(newFs.length, oldFs.length);
      for (let i = 0; i < maxLen; i++) {
        const newF = newFs[i];
        const oldF = oldFs[i];
        if (newF && !oldF) {
          findingChanges.push({ type: "new", column: newF.column, check: newF.check, newSeverity: severityName(newF.severity), message: newF.message });
        } else if (!newF && oldF) {
          findingChanges.push({ type: "resolved", column: oldF.column, check: oldF.check, oldSeverity: severityName(oldF.severity), message: oldF.message });
        } else if (newF && oldF) {
          if (newF.severity > oldF.severity) {
            findingChanges.push({ type: "worsened", column: newF.column, check: newF.check, oldSeverity: severityName(oldF.severity), newSeverity: severityName(newF.severity), message: newF.message });
          } else if (newF.severity < oldF.severity) {
            findingChanges.push({ type: "improved", column: newF.column, check: newF.check, oldSeverity: severityName(oldF.severity), newSeverity: severityName(newF.severity), message: newF.message });
          }
        }
      }
    }
  }

  for (const [key, oldFs] of oldGroups) {
    if (!newGroups.has(key)) {
      for (const oldF of oldFs) {
        findingChanges.push({
          type: "resolved",
          column: oldF.column,
          check: oldF.check,
          oldSeverity: severityName(oldF.severity),
          message: oldF.message,
        });
      }
    }
  }

  // Stat changes
  const rowDelta = newData.rowCount - oldData.rowCount;
  if (rowDelta !== 0) {
    statChanges.push({ metric: "row_count", oldValue: oldData.rowCount, newValue: newData.rowCount, delta: rowDelta });
  }
  const colDelta = newData.columns.length - oldData.columns.length;
  if (colDelta !== 0) {
    statChanges.push({ metric: "column_count", oldValue: oldData.columns.length, newValue: newData.columns.length, delta: colDelta });
  }

  return { schemaChanges, findingChanges, statChanges };
}

function severityName(s: Severity): string {
  switch (s) {
    case Severity.ERROR: return "ERROR";
    case Severity.WARNING: return "WARNING";
    case Severity.INFO: return "INFO";
  }
}

/**
 * Format a DiffReport as a human-readable string.
 */
export function formatDiffReport(report: DiffReport): string {
  const lines: string[] = [];

  if (report.schemaChanges.length > 0) {
    lines.push("Schema Changes:");
    for (const c of report.schemaChanges) {
      if (c.type === "added") lines.push(`  + ${c.column} (${c.newType})`);
      else if (c.type === "removed") lines.push(`  - ${c.column} (${c.oldType})`);
      else lines.push(`  ~ ${c.column}: ${c.oldType} → ${c.newType}`);
    }
  }

  if (report.findingChanges.length > 0) {
    lines.push("Finding Changes:");
    for (const c of report.findingChanges) {
      const prefix = c.type === "new" ? "+" : c.type === "resolved" ? "-" : "~";
      lines.push(`  ${prefix} [${c.type}] ${c.column}/${c.check}: ${c.message}`);
    }
  }

  if (report.statChanges.length > 0) {
    lines.push("Stat Changes:");
    for (const c of report.statChanges) {
      const sign = c.delta > 0 ? "+" : "";
      lines.push(`  ${c.metric}: ${c.oldValue} → ${c.newValue} (${sign}${c.delta})`);
    }
  }

  return lines.join("\n");
}
